src/utils/headers.js

```diff
@@ -3,45 +3,117 @@ const fs = require('fs')
-      if (index === rulePathParts.length - 1) {
-        return index === targetPathParts.length - 1
-      }
-      if (index === targetPathParts.length - 1) {
-        return false
-      }
-    } else {
-      return rulePart === target
+/**
+ * @param {Object!} rule
+ * The pattern to test.
+ *
+ * @param {string!} path
+ * The target path to match against.
+ *
+ * @returns {boolean}
+ * Whether or not the `rulePath` matches the `path`.
+ */
+const matchPaths = function (rule, path) {
+  /**
+   * Break the rule and target paths into pieces.
+   *
+   * /a/b/c -> ['a', 'b', 'c']
+   */
+  const ruleParts = rule.split('/').filter(Boolean)
+  const pathParts = path.split('/').filter(Boolean)
+  /**
+   * Fast path for /*.
+   */
+  const matchAll = ruleParts.length === 1 && ruleParts[0] === '*'
+  if (matchAll && path) return true
+  /**
+   * If either set of path parts is empty (indicating root dir /), they must
+   * both be empty (/ and /), otherwise there is no match.
+   */
+  const noRuleParts = ruleParts.length === 0
+  const noPathParts = pathParts.length === 0
+  if (noRuleParts || noPathParts) return noRuleParts && noPathParts
+  /**
+   * Otherwise, iterate over ruleParts and pathParts.
+   */
+  for (let index = 0; index < ruleParts.length; index++) {
+    /**
+     * Select the corresponding pathPart and rulePart.
+     */
+    const rulePart = ruleParts[index]
+    const pathPart = pathParts[index]
+    /**
+     * Whether or not all ruleParts and pathParts have been iterated over.
+     */
+    const noMoreRuleParts = index >= ruleParts.length - 1
+    const noMorePathParts = index >= pathParts.length - 1
+    /**
+     * What kind of placeholder, if any, the current rulePart is.
+     */
+    const ruleIsAsterisk = rulePart === '*'
+    const ruleIsPlaceholder = rulePart.startsWith(':')
+    const ruleIsWildcard = ruleIsAsterisk || ruleIsPlaceholder
+    /**
+     * For either either wildcard - (*) or (:placeholder) - the entire path is a
+     * definite match on this wildcard token if:
+     */
+    if (ruleIsWildcard) {
+      /**
+       * 1. the rulePart is a (*) wildcard AND it is the final rulePart, e.g.
+       *    `/path/to/*` matched against `/path/to/multiple/subdirs`.
+       */
+      if (noMoreRuleParts && ruleIsAsterisk) return true
+      /**
+       * 2. all ruleParts and pathParts have been iterated over, e.g.
+       *    `/path/to/:placeholder` and `/path/to/*` matched against
+       *    `/path/to/something`, as well as `/static/*` and
+       *    `/static/:placeholder` matched against `/static`)
+       */
+      if (noMorePathParts && noMoreRuleParts) return true
+    } else if (rulePart !== pathPart) {
+      /**
+       * If a mismatch is found, the rule does not match.
+       */
+      return false
+    } else if (noMoreRuleParts) {
+      /**
+       * If we have made it through all of the rules without finding a mismatch,
+       * the rule matches the path.
+       */
+      return true
```

Comment:
Refactoring `matchPaths` was the bulk of the work here. Be wary of `/path/to/dir` matching `/path/to/dir/:placeholder` under this logic (though `/path/:placeholder` will not match `/path/to/dir`).